### DISCUSSION TITLE
Testing: click is undefined

### DISCUSSION THREAD

**Discussion by Gw3i** [neuefische/questions#70] [Web] [answered]
Hello again,
I've an issue with my testing. I want to test if the error message "This field is required" will be shown when clicking on the submit button.
Unfortunately the test shows an error that ```click``` is ```undefined```.
Can someone see the problem?

Deployment:  https://capstone-project-git-videoform-gw3i.vercel.app/create
Sourcecode:  https://github.com/Gw3i/capstone-project/tree/videoForm

Error message:
```shell
FAIL  src/components/VideoForm.spec.js
  ● VideoForm › should show error message when input empty

    TypeError: Cannot read properties of undefined (reading 'click')

      17 |
      18 | 		const button = screen.getByRole('button', { name: /submit/i });
    > 19 | 		await userEvent.click(button);
         | 		                ^
      20 |
      21 | 		const errorMessageAfterClick = screen.getByText(/his field is required/i);
      22 | 		expect(errorMessageAfterClick).toBeInTheDocument();

      at Object.click (src/components/VideoForm.spec.js:19:19)

 PASS  src/components/Button/Button.spec.js
 PASS  src/utils/unit.spec.js

Test Suites: 1 failed, 2 passed, 3 total
Tests:       1 failed, 4 passed, 5 total
Snapshots:   0 total
Time:        1.124 s
```

**Comment by PReichetanz** [comment, marked as answer]
Hi Vladyslav,
the answer to your question is that the import of the `userEvent`has to be without curly braces:
`import userEvent from '@testing-library/user-event';`

However, if you do so, the test is still not working because `userEvent` is a little bit tricky.

I had to fix two typos and add a role attribute in your VideoForm.jsx:

```js
...
   <StyledInput
	{...register('YouTubeLink', {
	required: 'This field is required',
        // I changed this typo twice for each StyledInput
        ...
       <StyledInputWarning key={type} role="alert">
       // Add a role="alert" attribute to both StyledInputWarnings you can search for
	   {message}
      </StyledInputWarning>
```

After doing this, we can adjust the VideoForm.spec.js:

```js
import '@testing-library/jest-dom';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import VideoForm from './VideoForm';

describe('VideoForm', () => {
	it('should show error message when input empty', async () => {
		const user = userEvent.setup();
		render(<VideoForm />);

		const errorMessage = screen.queryByText(/This field is required/i);
		expect(errorMessage).not.toBeInTheDocument();

		const button = screen.getByRole('button', { name: /submit/i });
		await user.click(button);

		const errorMessageAfterClick = await waitFor(() => screen.getAllByRole('alert'));
		expect(errorMessageAfterClick).toHaveLength(2);
	});
});
```
- Note the `waitFor` and changed `userEvent` import.
- The Testing Library states to invoke `userEvent.setup()` [before the component is rendered](https://testing-library.com/docs/user-event/intro/#writing-tests-with-userevent).
- We have to wait for the rendering of the InputWarnings after the click of the submit button, this is why I used `waitFor()`.
- Since `errorMessageAfterClick` is now an array of two, we can check for the length; maybe you should rename this variable because it's not one message inside any more ;) 

Hope this helps!

**Reply by Gw3i** [reply]
This is great! Thank you very much, it helped me a lot 🙌
